truffle/src/main/java/org/jruby/truffle/core/rubinius/ByteArrayNodes.java

```diff
@@ -116,13 +217,46 @@ public Object getByte(DynamicObject bytes, DynamicObject pattern, int start, int
 
     }
 
+    @CoreMethod(names = "truncate", required = 1, lowerFixnum = 1)
+    public abstract static class TruncateNode extends CoreMethodArrayArgumentsNode {
+
+        @Specialization
+        public DynamicObject index(DynamicObject byteArray, int index) {
+            final ByteList byteList = Layouts.BYTE_ARRAY.getBytes(byteArray);
+            byteList.length(index);
+
+            return byteArray;
+        }
+    }
+
     @CoreMethod(names = "allocate", constructor = true)
     public abstract static class AllocateNode extends UnaryCoreMethodNode {
 
-        @TruffleBoundary
+        @Child private AllocateObjectNode allocateObjectNode;
+
+        public AllocateNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+            allocateObjectNode = AllocateObjectNode.create();
+        }
+
         @Specialization
         public DynamicObject allocate(DynamicObject rubyClass) {
-            throw new RaiseException(coreExceptions().typeErrorAllocatorUndefinedFor(rubyClass, this));
+            return allocateObjectNode.allocate(rubyClass, ByteList.EMPTY_BYTELIST);
```

Comment:
You could define `ByteArray.new` so allocation and initialization would not need to be split. `@Nullable` seems fine too.